zkevm-circuits/src/evm_circuit/execution/blockhash.rs

```diff
@@ -0,0 +1,201 @@
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        let block_number = block.rws[step.rw_indices[0]].stack_value();
+        self.block_number.assign(
+            region,
+            offset,
+            Some(
+                block_number.to_le_bytes()[..N_BYTES_U64]
+                    .try_into()
+                    .unwrap(),
+            ),
+        )?;
+        let block_number: F = block_number.to_scalar().unwrap();
+
+        let current_block_number = block.context.number;
+        self.current_block_number
+            .assign(region, offset, current_block_number.to_scalar())?;
+
+        self.block_hash.assign(
+            region,
+            offset,
+            Some(block.rws[step.rw_indices[1]].stack_value().to_le_bytes()),
+        )?;
+
+        self.block_lt.assign(
+            region,
+            offset,
+            block_number,
+            current_block_number.to_scalar().unwrap(),
+        )?;
+
+        self.diff_lt.assign(
+            region,
+            offset,
+            current_block_number.to_scalar().unwrap(),
+            block_number + F::from(256),
+        )?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{
+        evm_circuit::witness::block_convert,
+        test_util::{test_circuits_using_witness_block, BytecodeTestConfig},
+    };
+    use bus_mapping::mock::BlockData;
+    use eth_types::{bytecode, geth_types::GethData, Bytecode, U256};
+    use mock::test_ctx::{helpers::*, TestContext};
+
+    fn test_ok(block_number: usize, current_block_number: u64) {
+        let mut code = Bytecode::default();
+        code.append(&bytecode! {
+            PUSH32(block_number)
+            BLOCKHASH
+            STOP
+        });
+
+        // simple U256 values for history hashes
+        let mut history_hashes = Vec::new();
+        if current_block_number < 256 {
+            for number in 0..current_block_number {
+                history_hashes.push(U256::from(number));
+            }
+        } else {
+            for number in (current_block_number - 256)..current_block_number {
+                history_hashes.push(U256::from(number));
+            }
+        }
+        let block: GethData = TestContext::<2, 1>::new(
+            Some(history_hashes),
+            account_0_code_account_1_no_code(code),
+            tx_from_1_to_0,
+            |block, _tx| block.number(current_block_number),
+        )
+        .unwrap()
+        .into();
+
+        let mut builder = BlockData::new_from_geth_data(block.clone()).new_circuit_input_builder();
+        builder
+            .handle_block(&block.eth_block, &block.geth_traces)
+            .expect("could not handle block tx");
+
+        test_circuits_using_witness_block(
+            block_convert(&builder.block, &builder.code_db),
+            BytecodeTestConfig::default(),
+        )
+        .unwrap();
+    }
+    #[test]
+    fn blockhash_gadget_test() {
+        test_ok(6, 5);
```

Comment:
Shouldn't this fail? How are you suposed to check the block hash of block 6 if the current one is 5?

Reply:
If `BLOCKHASH` got input from invalid range, it pushes 0 to the stack. So Imo, that case should not fail. Isn't it?

Reply:
You're right. My bad. :smile: